docs/notebooks/trust_region.pct.py

```diff
@@ -0,0 +1,286 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: -all
+#     custom_cell_magics: kql
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.2
+#   kernelspec:
+#     display_name: .venv_310
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Trust region Bayesian optimization
+#
+# We will demonstrate three trust region Bayesian optimization algorithms in this tutorial.
+
+# %%
+import numpy as np
+import tensorflow as tf
+
+np.random.seed(1793)
+tf.random.set_seed(1793)
+
+# %% [markdown]
+# ## Define the problem and model
+#
+# We can use trust regions for Bayesian optimization in much the same way as we used EGO and EI in
+# the [introduction notebook](expected_improvement.ipynb). Since the setup is very similar to
+# that tutorial, we'll skip over most of the detail.
+
+# %%
+import trieste
+from trieste.objectives import Branin
+
+branin = Branin.objective
+search_space = Branin.search_space
+
+num_initial_data_points = 10
+initial_query_points = search_space.sample(num_initial_data_points)
+observer = trieste.objectives.utils.mk_observer(branin)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# As usual, we'll use Gaussian process regression to model the function. Note that we set the
+# likelihood variance to a small number because we are dealing with a noise-free problem.
+
+# %%
+from trieste.models.gpflow import GaussianProcessRegression, build_gpr
+
+
+def build_model():
+    gpflow_model = build_gpr(
+        initial_data, search_space, likelihood_variance=1e-7
+    )
+    return GaussianProcessRegression(gpflow_model)
+
+
+# %% [markdown]
+# ## Trust region `TREGO` acquisition rule
+#
+# First we show how to run Bayesian optimization with the `TREGO` algorithm. This is a trust region
+# algorithm that alternates between regular EGO steps and local steps within one trust region.
+#
+# ### Create `TREGO` rule and run optimization loop
+#
+# We can run the Bayesian optimization loop by defining a `BayesianOptimizer` and calling its
+# `optimize` method with the trust region rule. Once the optimization loop is complete, the
+# optimizer will return one new query point for every step in the loop; that's 5 points in total.
+
+# %%
+acq_rule = trieste.acquisition.rule.TrustRegion()
```

Comment:
Possibly it's important here to explain that TREGO is "meta" rule: it applies a base rule, either inside a Trust Region, or over the whole space. I guess the default is to use EfficientGlobalOptimization if nothing is specified but it would be good to make it explicit here? Would also show users what they need to change if they want to use a custom rule?

Reply:
Yes EGO is the default. I have made that explicit now, with an explanation.